src/core/bloom.cc

```diff
@@ -0,0 +1,105 @@
+// Copyright 2024, DragonflyDB authors.  All rights reserved.
+// See LICENSE for licensing terms.
+//
+
+#include "core/bloom.h"
+
+#include <absl/base/internal/endian.h>
+#include <absl/numeric/bits.h>
+#include <mimalloc.h>
+
+#include <cmath>
+
+#define XXH_STATIC_LINKING_ONLY
+#include <xxhash.h>
+
+#include "base/logging.h"
+
+namespace dfly {
+
+using namespace std;
+
+namespace {
+
+inline XXH128_hash_t Hash(string_view str) {
+  return XXH3_128bits_withSeed(str.data(), str.size(), 0xc6a4a7935bd1e995ULL);  // murmur2 seed
+}
+
+inline uint64_t BitIndex(const XXH128_hash_t& hash, unsigned i, uint64_t mask) {
+  return (hash.low64 + hash.high64 * i) % mask;
+}
+
+}  // namespace
+
+Bloom::Bloom(uint32_t entries, double error, mi_heap_t* heap) {
+  CHECK(error > 0 && error < 1);
+
+  if (entries < 1024)
+    entries = 1024;
+
+  constexpr double kDenom = M_LN2 * M_LN2;
+  double bpe = -log(error) / kDenom;
+
+  hash_cnt_ = ceil(M_LN2 * bpe);
+
+  uint64_t bits = uint64_t(ceil(entries * bpe));
+  bits = absl::bit_ceil(bits);  // make it power of 2.
+  if (bits < 1024) {
+    bits = 1024;
+  }
+
+  uint64_t length = bits / 8;
+
+  bf_ = (uint8_t*)mi_heap_calloc(heap, length, 1);
+
```

Comment:
just a thought, but an utuility that returns a unique_ptr on the mi-heap with a custom deleter would be helpful, we do this currently for some pointers in connection

Reply:
Maybe but then we will increase the class size